django/core/servers/basehttp.py

```diff
@@ -227,13 +227,21 @@ def handle_one_request(self):
         handler.run(self.server.get_app())
 
 
-def run(addr, port, wsgi_handler, ipv6=False, threading=False, server_cls=WSGIServer):
+def run(
+    addr,
+    port,
+    wsgi_handler,
+    ipv6=False,
+    threading=False,
+    on_bind=None,
+    server_cls=WSGIServer,
+):
     server_address = (addr, port)
-    if threading:
-        httpd_cls = type("WSGIServer", (socketserver.ThreadingMixIn, server_cls), {})
-    else:
-        httpd_cls = server_cls
+    bases = (socketserver.ThreadingMixIn, server_cls) if threading else (server_cls,)
+    httpd_cls = type("WSGIServer", bases, {})
     httpd = httpd_cls(server_address, WSGIRequestHandler, ipv6=ipv6)
+    if on_bind is not None:
+        on_bind(httpd.server_port)
```

Comment:
Btw if we are afraid that some `server_cls`  implementations might not expose `server_port/server_name` we could hook into the base environ where this needs to be set I think (though it is questionable if that is exposed under the same name)

Reply:
Can we use `port` passed to `run()` as a fallback?
```suggestion
        on_bind(getattr(httpd, "server_port", port))
```